handler-utils.go

```diff
@@ -99,32 +98,33 @@ func extractMetadataFromHeader(header http.Header) map[string]string {
 	return metadata
 }
 
-func extractHTTPFormValues(reader *multipart.Reader) (io.Reader, string, map[string]string, error) {
+// Extract form fields and file data from a HTTP POST Policy
+func extractPostPolicyFormValues(reader *multipart.Reader) (filePart io.Reader, fileName string, formValues map[string]string, err error) {
 	/// HTML Form values
-	formValues := make(map[string]string)
-	filePart := new(bytes.Buffer)
-	fileName := ""
-	var err error
+	formValues = make(map[string]string)
+	fileName = ""
 	for err == nil {
 		var part *multipart.Part
 		part, err = reader.NextPart()
 		if part != nil {
 			canonicalFormName := http.CanonicalHeaderKey(part.FormName())
 			if canonicalFormName != "File" {
 				var buffer []byte
-				buffer, err = ioutil.ReadAll(part)
+				// Limit fields size (except file) to 1Mib since Policy document
+				// can reach that size according to https://aws.amazon.com/articles/1434
+				limitReader := io.LimitReader(part, 1024*1024)
```

Comment:
Also do 1024*1024+1 - so that ioutil.ReadAll() returns a buffer we can validate if it is bigger than 1Mib and throw an appropriate error as well.